stablehlo/dialect/CheckOps.td

```diff
@@ -0,0 +1,132 @@
+/* Copyright 2023 The TensorFlow Authors. All Rights Reserved.
+   Copyright 2023 The StableHLO Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef STABLEHLO_DIALECT_CHECK_OPS
+#define STABLEHLO_DIALECT_CHECK_OPS
+
+include "stablehlo/dialect/Base.td"
+
+//===----------------------------------------------------------------------===//
+// CHECK dialect.
+//===----------------------------------------------------------------------===//
+
+def CHECK_Dialect : Dialect {
```

Comment:
Given that this dialect is specific to tests, I think it should live under `tests/`.

Reply:
done

Reply:
Do you think `tests/` even though interpreter needs to support these ops? Currently only testing utils and tools depend on any files in our `tests/` folder. Something feels a bit off about `reference/` depending on something from  `test/`

Reply:
"Something feels a bit off about reference/ depending on something from tests/". That's a good point. Continuing that logic, would you agree that it's odd for `reference/` to include stuff related to testing, e.g. for Ops.h to include `evalEqOp`?

Reply:
From a user perspective - do we ever intend another project to take a dependency on a reference library? If so would we expect them to want check ops? If yes, I would think that verification (especially the machinery which determines an acceptable threshold for FP equality) is a feature of a reference interpreter.

If not, then yeah I agree this is test machinery specific to StableHLO testing and a perfect design probably would allow plugging custom dialects into the reference interpreter, like these test ops. That feels like overkill, so the current design is probably good enough for now. I think that's what IREE does? It has some VM Native implementation declaration.

Reply:
"a perfect design probably would allow plugging custom dialects into the reference interpreter, like these test ops". A low-tech alternative to decouple the check dialect, if we decide this is desirable, would involve splitting `--interpret` into `--interpret` (reference-specific) and `--check` (tests-specific).

Reply:
Hi All, Thanks for all the comments. 

>> do we ever intend another project to take a dependency on a reference library?

Thanks for the feedback! 
I do not think the consumers of StableHLO op `eval`uators should be interested in `Check` specific ops. Using  `checkDialect` is one way for testing the evaluators and it is not unlikely for a consumer to have different testing strategies which may not reply on those. One such need could be to cross validate their implementation of StableHLO ops  against the interpreter.


Having said that this is my proposal:
## Proposal 
1. Move the definition of `Check` ops in test/.
2. Move the reference implementation of `Check` ops in test/
3. Move `Interpreter.cpp` in tools/.

With the above, the binary for `stablehlo-interpreter` will evolved as follows:
B1. dialect/ and refernece/ respectively provide the libraries related to  StableHLO ops and  the implementation of those ops.
B2. test/ provides the the libraries related to  both CheckOps ops and  the implementation of those ops.
B3. (1) and (2) contribute to create a library w.r.t tools/Interpreter.cpp (which is the interpreter driver for different dialects; more about this later)
B4. Finally (3) is linked to create the executable at tool/.

### Rationale for 1 and 2
Both the dialect and its implementation are need for testing purposes only.  

### rationale for 3
The  `Interpreter.cpp` provides the evaluator for `func::func` op. Specifically, it provides a an interpreter drive loop to evaluate op from different dialects (like `func::retrunOp, Stablehlo::*, check::*`). IMO, it should not be part of reference/ which host the evaluator of StableHLO ops only.

To me "..nterpreter drive loop to evaluate op from different dialects.." seems very close to tooling. making  
`Interpreter.cpp` part of test/ has the problem that consumer has to depend on test/ unless they are willing to write their own driver.

### Outcome
reference/ has now **no** dependency on tests.  This enables writing an interpreter driver depending solely on reference/.

## About the added dependency from from test/ to tool/  (because of B4)

Given that we already have similar  dependency w.r.t `StablehloTestUtils`, at least we are no introducing a new dependency. However, we can provide additional machinery: For example two interpreter tools like `check-stablehlo-interpreter` (depending on check/,  dialect/ and reference), and  `stablehlo-interpreter` (depending on only dialect/ and reference). For our use case only the former makes sense.



>>  A low-tech alternative to decouple the check dialect, if we decide this is desirable, would involve splitting --interpret into --interpret (reference-specific) and --check (tests-specific).
This sound great! Given the above proposal, do we need to explore this? @burmako  

Let me know your feedback.

Reply:
> IMO, it should not be part of reference/ which host the evaluator of StableHLO ops only 

I think that the reference implementation should cover all ops that constitute StableHLO programs, because that's the fundamental use case for the reference implementation - take a StableHLO program and run it.

Today, some of these ops come from other dialects (e.g. builtin/func for modularity, arith/shape/tensor for dynamism). In the future, we are planning to fork most of these ops into StableHLO. But regardless of where these ops live, I think that we want to handle them in the interpreter (i.e. under `reference/`).

> Move Interpreter.cpp in tools/ (or tests/ with trade offs discussed later 

If we abstract away from naming a little bit (whether the file is called Interpreter.h or something else) and view this from the perspective of the functionality that we want `reference/` to provide, I think this functionality would look roughly as follows:

```C++
EvaluationResult evaluate(StablehloProgram program, ArrayRef<Argument> args);
```

"making Interpreter.cpp part of test/ has the problem that consumer has to depend on test/ unless they are willing to write their own driver". I would reframe this as "has the problem that consumer will need to implement the functionality that they need on their own, using the individual `evalFoo` functions provided by `reference/`". When viewed from this angle, I think this would be a considerable downside - it would be quite undesirable for us to leave parts of the implementation of our core functionality to users.

Given that, I would recommend to find a way to keep Interpreter.h under `reference/`, e.g. by changing `eval` to take an optional parameter callback that looks along the lines of `llvm::function_ref<void(Operation*, Scope*)>`. Then regular users won't even notice this, and the `stablehlo-interpreter` tool would be able to pass a callback that handles `check` ops.

Reply:
Thanks for the suggestion. It looks great to me. Implemented, PTAL.